compiler/src/dotty/tools/dotc/typer/Implicits.scala

```diff
@@ -892,106 +886,158 @@ trait Implicits { self: Typer =>
    */
   lazy val synthesizedProductMirror: SpecialHandler =
     (formal: Type, span: Span) => implicit (ctx: Context) => {
-      def mirrorFor(monoType: Type): Tree = monoType match {
-        case AndType(tp1, tp2) =>
-          mirrorFor(tp1).orElse(mirrorFor(tp2))
-        case _ =>
-          if (monoType.termSymbol.is(CaseVal)) {
-            val module = monoType.termSymbol
-            val modulePath = pathFor(monoType).withSpan(span)
-            if (module.info.classSymbol.is(Scala2x)) {
-              val mirrorType = mirrorCore(defn.Mirror_SingletonProxyType, monoType, module.name)
-              val mirrorRef = New(defn.Mirror_SingletonProxyType, modulePath :: Nil)
-              mirrorRef.cast(mirrorType)
+      def mirrorFor(mirroredType0: Type): Tree = {
+        val mirroredType = mirroredType0.stripTypeVar
+        mirroredType match {
+          case AndType(tp1, tp2) =>
+            mirrorFor(tp1).orElse(mirrorFor(tp2))
+          case _ =>
+            if (mirroredType.termSymbol.is(CaseVal)) {
+              val module = mirroredType.termSymbol
+              val modulePath = pathFor(mirroredType).withSpan(span)
+              if (module.info.classSymbol.is(Scala2x)) {
+                val mirrorType = mirrorCore(defn.Mirror_SingletonProxyType, mirroredType, mirroredType, module.name)
+                val mirrorRef = New(defn.Mirror_SingletonProxyType, modulePath :: Nil)
+                mirrorRef.cast(mirrorType)
+              }
+              else {
+                val mirrorType = mirrorCore(defn.Mirror_SingletonType, mirroredType, mirroredType, module.name)
+                modulePath.cast(mirrorType)
+              }
             }
-            else {
-              val mirrorType = mirrorCore(defn.Mirror_SingletonType, monoType, module.name)
-              modulePath.cast(mirrorType)
+            else if (mirroredType.classSymbol.isGenericProduct) {
+              val cls = mirroredType.classSymbol
+              val accessors = cls.caseAccessors.filterNot(_.is(PrivateLocal))
+              val elemLabels = accessors.map(acc => ConstantType(Constant(acc.name.toString)))
+              val (monoType, elemTypes) = mirroredType match {
+                case mirroredType: HKTypeLambda =>
+                  val elems =
```

Comment:
I'd rename to `elemsType` for clarity (in both branches). I used `elems` as a cue that this was a list of types, and then got confused.